tools/clang/lib/SPIRV/SPIRVEmitter.cpp

```diff
@@ -992,10 +1005,62 @@ void SPIRVEmitter::doReturnStmt(const ReturnStmt *stmt) {
   }
 }
 
+bool SPIRVEmitter::isBreakStmtLastStmtInCaseStmt(const BreakStmt *breakStmt,
+                                                 const SwitchStmt *switchStmt) {
+  std::vector<const Stmt *> flatSwitch;
+  flattenSwitchStmtAST(switchStmt->getBody(), &flatSwitch);
+  auto iter =
+      std::find(flatSwitch.begin(), flatSwitch.end(), cast<Stmt>(breakStmt));
+  assert(iter != std::end(flatSwitch));
+
+  // We are interested to know what comes after the BreakStmt.
+  ++iter;
+
+  // The break statement is the last statement in its case statement iff:
+  // it is the last statement in the switch, or,
+  // its next statement is either 'default' or 'case'
+  return iter == flatSwitch.end() || isa<CaseStmt>(*iter) ||
+         isa<DefaultStmt>(*iter);
+}
+
+const Stmt* SPIRVEmitter::breakStmtScope(const BreakStmt *breakStmt) {
+  const Stmt *curNode = nullptr;
+  for (curNode = getImmediateParent(astContext, breakStmt);
+       curNode && !isLoopStmt(curNode) && !isa<SwitchStmt>(curNode);
+       curNode = getImmediateParent(astContext, curNode))
+    ;
+  return curNode;
+}
+
 void SPIRVEmitter::doBreakStmt(const BreakStmt *breakStmt) {
+  assert(!theBuilder.isCurrentBasicBlockTerminated());
   uint32_t breakTargetBB = breakStack.top();
   theBuilder.addSuccessor(breakTargetBB);
   theBuilder.createBranch(breakTargetBB);
+
+  // If any statements follow a break statement in a loop or switch, they will
+  // not be executed. For example, StmtB and StmtC below are never executed:
+  //
+  // while (true) {
+  //   StmtA;
+  //   break;
+  //   StmtB;
+  //   StmtC;
+  // }
+  //
+  // To handle such cases, we do not stop tranlsation. We create a new basic
+  // block in which StmtB and StmtC will be translated.
+  // Note that since this basic block is unreachable, BlockReadableOrderVisitor
+  // will not emit it in the final module binary.
+  const Stmt *scope = breakStmtScope(breakStmt);
+  if ((isa<SwitchStmt>(scope) && !isBreakStmtLastStmtInCaseStmt(
```

Comment:
Minor: I'd suggest to push some comments inside the `if` check since it's quite complicated:
```c++
if (
  // Have unreachable instructions after a break statement in a case of a switch statement
  (isa<SwitchStmt>(scope) && !..) ||
  // Have unreachable instructions after a break statement in a loop statement
  (isLoopStmt(scope) && !...))
```